datalad/nonasyncrunner.py

```diff
@@ -154,6 +165,11 @@ def run_command(cmd: Union[str, List],
         process_stderr_fileno: STDERR_FILENO
     }
 
+    if input is not None:
+        # (ab)use internal helper that takes care of a bunch of corner cases
+        # and closes stdin at the end
+        process._stdin_write(input)
+
```

Comment:
I think we should probably do that after the output reader threads are started, otherwise we might end up with a deadlock: subprocess blocked on stdout/stderr, and, as a result of that, a blocked main thread, if input is large and the input-fifo runs full. That might only happen with inputs larger than 4k, but we do not restrict the size, as far as I know.

Still, it might also be good to do the writing in an extra thread, since the main thread should read the output queues to ensure that the sub-process does not block on its output.

I would propose to introduce a writer thread (or re-introduce it, it existed in a previous non-async-commit) to ensure that we do not end up with blocked subprocesses or deadlocks.

Reply:
I tried that (without a dedicated thread), but it did not work. If you see how this could be done, please take over. Thx!

PS: I am already using this feature in the implementation of the git-credential support.

Reply:
Will give it a try, and get back to you.